<a href="https://colab.research.google.com/github/ahmedaouadi91/food-sales-predictions/blob/main/Project_1_Final_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns',100)

In [ ]:
fpath = '/content/drive/MyDrive/sales_predictions_2023.csv'
df = pd.read_csv(fpath)
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [ ]:
#data type of each variable
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [ ]:
#duplicated rows
df.duplicated().sum()

0

In [ ]:
#dropping 'Item_Identifier' column
df.drop(columns='Item_Identifier', inplace=True)
df.head()

Item_Weight Item_Fat_Content  Item_Visibility              Item_Type  \
0         9.30          Low Fat         0.016047                  Dairy   
1         5.92          Regular         0.019278            Soft Drinks   
2        17.50          Low Fat         0.016760                   Meat   
3        19.20          Regular         0.000000  Fruits and Vegetables   
4         8.93          Low Fat         0.000000              Household   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  249.8092            OUT049                       1999      Medium   
1   48.2692            OUT018                       2009      Medium   
2  141.6180            OUT049                       1999      Medium   
3  182.0950            OUT010                       1998         NaN   
4   53.8614            OUT013                       1987        High   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0               Tier 1  Supermarket Type1          3735.1380  
1               Tier 3  Supermarket Type2           443.4228  
2               Tier 1  Supermarket Type1          2097.2700  
3               Tier 3      Grocery Store           732.3800  
4               Tier 3  Supermarket Type1           994.7052

To find and fix any inconsistent categories of data i will fix columns by columns
* Object

In [ ]:
col_obj = df.select_dtypes('object').columns
col_obj

Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [ ]:
#Item_Fat_Content
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [ ]:
df['Item_Fat_Content']=df['Item_Fat_Content'].replace('LF','Low Fat')
df['Item_Fat_Content']=df['Item_Fat_Content'].replace('reg','Regular')
df['Item_Fat_Content']=df['Item_Fat_Content'].replace('low fat','Low Fat')
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [ ]:
#Item_Type
df['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

In [ ]:
#Outlet_Size
df['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [ ]:
#Outlet_Location_Type
df['Outlet_Location_Type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

In [ ]:
#Outlet_Type
df['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

* Numeric

In [ ]:
col_num = df.select_dtypes('number').columns
col_num

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year', 'Item_Outlet_Sales'],
      dtype='object')

In [ ]:
df.describe().loc[['min','max','mean']].round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
min          4.56             0.00     31.29                    1985.00   
max         21.35             0.33    266.89                    2009.00   
mean        12.86             0.07    140.99                    1997.83   

      Item_Outlet_Sales  
min               33.29  
max            13086.96  
mean            2181.29

dentify the features (X) and target (y)

In [ ]:
# The target we are trying to predict
y = df['Item_Outlet_Sales']
# The features we will use to make the prediction
X = df.drop(columns = 'Item_Outlet_Sales')

Perform a train test split

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
X_train.head()

Item_Weight Item_Fat_Content  Item_Visibility     Item_Type  Item_MRP  \
4776       16.350          Low Fat         0.029565     Household  256.4646   
7510       15.250          Regular         0.000000   Snack Foods  179.7660   
5828       12.350          Regular         0.158716          Meat  157.2946   
5327        7.975          Low Fat         0.014628  Baking Goods   82.3250   
4810       19.350          Low Fat         0.016645  Frozen Foods  120.9098   

     Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
4776            OUT018                       2009      Medium   
7510            OUT018                       2009      Medium   
5828            OUT049                       1999      Medium   
5327            OUT035                       2004       Small   
4810            OUT045                       2002         NaN   

     Outlet_Location_Type        Outlet_Type  
4776               Tier 3  Supermarket Type2  
7510               Tier 3  Supermarket Type2  
5828               Tier 1  Supermarket Type1  
5327               Tier 2  Supermarket Type1  
4810               Tier 2  Supermarket Type1

# **Create a preprocessing object to prepare the dataset for Machine Learning**

---



# A-Imputation of Numeric Features with Median

1) Create a list of features

In [ ]:
# Get the list of numeric column names
num_cols = X_train.select_dtypes("number").columns
num_cols

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')

2) Instantiate the Imputer

In [ ]:
from sklearn.impute import SimpleImputer

In [ ]:
# Instantiate the imputer object from the SimpleImputer class with strategy 'median'
impute_median = SimpleImputer(strategy='median')

3) Fit the Imputer of the Training Data

In [ ]:
# Fit the imputer object on the numeric training data with .fit()
impute_median.fit(X_train[num_cols])

SimpleImputer(strategy='median')

4a) Transform the Training Data

In [ ]:
from sklearn import set_config
set_config(transform_output='pandas')

In [ ]:
# Transform the training data (fill in nulls with median)
X_train_num_imputed = impute_median.transform(X_train[num_cols])
X_train_num_imputed

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year
4776       16.350         0.029565  256.4646                     2009.0
7510       15.250         0.000000  179.7660                     2009.0
5828       12.350         0.158716  157.2946                     1999.0
5327        7.975         0.014628   82.3250                     2004.0
4810       19.350         0.016645  120.9098                     2002.0
...           ...              ...       ...                        ...
5734        9.395         0.286345  139.1838                     1998.0
5191       15.600         0.117575   75.6670                     2007.0
5390       17.600         0.018944  237.3590                     2002.0
860        20.350         0.054363  117.9466                     2007.0
7270       16.350         0.016993   95.7410                     1997.0

[6392 rows x 4 columns]

In [ ]:
#Let's confirm that our transformed numeric training data no longer has null values.
X_train_num_imputed.isna().sum()

Item_Weight                  0
Item_Visibility              0
Item_MRP                     0
Outlet_Establishment_Year    0
dtype: int64

4b) Transform the Test Data

In [ ]:
# Transform the testing data (fill in nulls with median from training data)
X_test_num_imputed = impute_median.transform(X_test[num_cols])
X_test_num_imputed.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year
7503       14.300         0.026300   79.4302                     1987.0
2957        7.930         0.071136   42.7086                     1997.0
7031       14.500         0.041313   42.0454                     1999.0
1084       12.650         0.044767  173.7054                     1985.0
856        10.195         0.012456  197.5110                     2004.0

In [ ]:
#Let's confirm that our transformed numeric training data no longer has null values.
X_test_num_imputed.isna().sum()

Item_Weight                  0
Item_Visibility              0
Item_MRP                     0
Outlet_Establishment_Year    0
dtype: int64

# Imputation of Categorical Features

Define the List of Features

In [ ]:
col_obj

Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [ ]:
# Get the list of categorical column names
cat_exp=['Item_Fat_Content' , 'Item_Type', 'Outlet_Identifier', 'Outlet_Location_Type', 'Outlet_Type']
cat_cols = X_train[cat_exp].columns
cat_cols

Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

2) Instantiate the Imputer

In [ ]:
# Instantiate the new imputer
impute_na = SimpleImputer(strategy='constant', fill_value='NA')

3) Fit the Imputer on the Training Data

In [ ]:
# Fit the imputer on the training data
impute_na.fit(X_train[cat_cols])

SimpleImputer(fill_value='NA', strategy='constant')

4a) Transform the Training Data

In [ ]:
# Transform the categorical training data
X_train_cat_imputed = impute_na.transform(X_train[cat_cols])

In [ ]:
# Confirm that all null vales in cat columns are filled in training set
X_train_cat_imputed.isna().sum().sum()

0

4b) Transform the Testing Data

In [ ]:
# Transform the categorical testing data
X_test_cat_imputed = impute_na.transform(X_test[cat_cols])

In [ ]:
# Confirm all missing values in test set are filled
X_test_cat_imputed.isna().sum().sum()

0

# Imputation Ordinal Features

1) Define the List of Features

In [ ]:
ord_exp=['Outlet_Size']
ord_cols = X_train[ord_exp].columns
ord_cols

Index(['Outlet_Size'], dtype='object')

2) Instantiate the Imputer

In [ ]:
# Instantiate the new imputer
impute_na = SimpleImputer(strategy='constant', fill_value='NA')

3) Fit the Imputer on the Training Data

In [ ]:
# Fit the imputer on the training data
impute_na.fit(X_train[ord_cols])

SimpleImputer(fill_value='NA', strategy='constant')

4a) Transform the Training Data

In [ ]:
# Transform the ordinal training data
X_train_ord_imputed = impute_na.transform(X_train[ord_cols])

In [ ]:
# Confirm all missing values in train set are filled
X_train_ord_imputed.isna().sum().sum()

0

4a) Transform the Testing Data

In [ ]:
# Transform the ordinal training data
X_test_ord_imputed = impute_na.transform(X_test[ord_cols])

In [ ]:
# Confirm all missing values in test set are filled
X_test_ord_imputed.isna().sum().sum()

0

# Create a StandardScaler for scaling numeric columns.

In [ ]:
# New import for scaler
from sklearn.preprocessing import StandardScaler

In [ ]:
# instantiate scaler
scaler = StandardScaler()

In [ ]:
# fit scaler on training data
scaler.fit(X_train_num_imputed)

StandardScaler()

In [ ]:
# transform training data
X_train_num_scaled = scaler.transform(X_train_num_imputed)
# transform testing data
X_test_num_scaled = scaler.transform(X_test_num_imputed)

# Ordinal Encoding

In [ ]:
# New Imports
from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder

In [ ]:
# Specifying the order of categories in quality/condition columns
qual_cond_order = ['NA','Small', 'Medium', 'High']

In [ ]:
df['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [ ]:
ordinal_category_orders = [qual_cond_order]

In [ ]:
# Instantiate the encoder and include the list of ordered values as an argument
ord_encoder = OrdinalEncoder(categories=ordinal_category_orders)
ord_encoder

OrdinalEncoder(categories=[['NA', 'Small', 'Medium', 'High']])

In [ ]:
# Fit the encoder on the training data
ord_encoder.fit(X_train_ord_imputed[ord_cols])

OrdinalEncoder(categories=[['NA', 'Small', 'Medium', 'High']])

4a) Transform the training data

In [ ]:
# Transform the training data
X_train_ord_encoder = ord_encoder.transform(X_train_ord_imputed[ord_cols])

In [ ]:
# Transform the testing data
X_test_ord_encoder = ord_encoder.transform(X_test_ord_imputed[ord_cols])

# Categorical Encoding

In [ ]:
# saving list of categorical features to one-hot-encode
cat_exp=['Item_Fat_Content' , 'Item_Type', 'Outlet_Identifier', 'Outlet_Location_Type', 'Outlet_Type']
cat_cols = X_train[cat_exp].columns
cat_cols

Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [ ]:
# Instantiate one hot encoder
ohe_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

In [ ]:
# Fit the OneHotEncoder on the training data
ohe_encoder.fit(X_train_cat_imputed[cat_cols])

OneHotEncoder(handle_unknown='ignore', sparse_output=False)

In [ ]:
# Transform the training data
X_train_cat_ohe = ohe_encoder.transform(X_train_cat_imputed[cat_cols])
X_train_cat_ohe

Item_Fat_Content_Low Fat  Item_Fat_Content_Regular  \
4776                       1.0                       0.0   
7510                       0.0                       1.0   
5828                       0.0                       1.0   
5327                       1.0                       0.0   
4810                       1.0                       0.0   
...                        ...                       ...   
5734                       0.0                       1.0   
5191                       1.0                       0.0   
5390                       1.0                       0.0   
860                        1.0                       0.0   
7270                       1.0                       0.0   

      Item_Type_Baking Goods  Item_Type_Breads  Item_Type_Breakfast  \
4776                     0.0               0.0                  0.0   
7510                     0.0               0.0                  0.0   
5828                     0.0               0.0                  0.0   
5327                     1.0               0.0                  0.0   
4810                     0.0               0.0                  0.0   
...                      ...               ...                  ...   
5734                     0.0               0.0                  0.0   
5191                     0.0               0.0                  0.0   
5390                     0.0               0.0                  0.0   
860                      0.0               0.0                  0.0   
7270                     0.0               0.0                  0.0   

      Item_Type_Canned  Item_Type_Dairy  Item_Type_Frozen Foods  \
4776               0.0              0.0                     0.0   
7510               0.0              0.0                     0.0   
5828               0.0              0.0                     0.0   
5327               0.0              0.0                     0.0   
4810               0.0              0.0                     1.0   
...                ...              ...                     ...   
5734               0.0              0.0                     0.0   
5191               0.0              0.0                     1.0   
5390               0.0              0.0                     0.0   
860                0.0              0.0                     0.0   
7270               0.0              0.0                     0.0   

      Item_Type_Fruits and Vegetables  Item_Type_Hard Drinks  \
4776                              0.0                    0.0   
7510                              0.0                    0.0   
5828                              0.0                    0.0   
5327                              0.0                    0.0   
4810                              0.0                    0.0   
...                               ...                    ...   
5734                              1.0                    0.0   
5191                              0.0                    0.0   
5390                              0.0                    0.0   
860                               0.0                    0.0   
7270                              0.0                    0.0   

      Item_Type_Health and Hygiene  Item_Type_Household  Item_Type_Meat  \
4776                           0.0                  1.0             0.0   
7510                           0.0                  0.0             0.0   
5828                           0.0                  0.0             1.0   
5327                           0.0                  0.0             0.0   
4810                           0.0                  0.0             0.0   
...                            ...                  ...             ...   
5734                           0.0                  0.0             0.0   
5191                           0.0                  0.0             0.0   
5390                           1.0                  0.0             0.0   
860                            0.0                  0.0             0.0   
7270                           0.

In [ ]:
# Transform the testing data
X_test_cat_ohe = ohe_encoder.transform(X_test_cat_imputed[cat_cols])
X_test_cat_ohe

Item_Fat_Content_Low Fat  Item_Fat_Content_Regular  \
7503                       1.0                       0.0   
2957                       1.0                       0.0   
7031                       0.0                       1.0   
1084                       0.0                       1.0   
856                        0.0                       1.0   
...                        ...                       ...   
4753                       1.0                       0.0   
4836                       1.0                       0.0   
8064                       1.0                       0.0   
4418                       1.0                       0.0   
1784                       1.0                       0.0   

      Item_Type_Baking Goods  Item_Type_Breads  Item_Type_Breakfast  \
7503                     0.0               0.0                  0.0   
2957                     0.0               0.0                  0.0   
7031                     0.0               0.0                  0.0   
1084                     0.0               0.0                  0.0   
856                      0.0               0.0                  0.0   
...                      ...               ...                  ...   
4753                     0.0               0.0                  0.0   
4836                     0.0               0.0                  0.0   
8064                     0.0               0.0                  0.0   
4418                     0.0               0.0                  0.0   
1784                     0.0               0.0                  0.0   

      Item_Type_Canned  Item_Type_Dairy  Item_Type_Frozen Foods  \
7503               0.0              0.0                     1.0   
2957               0.0              0.0                     0.0   
7031               1.0              0.0                     0.0   
1084               0.0              0.0                     0.0   
856                0.0              0.0                     0.0   
...                ...              ...                     ...   
4753               0.0              1.0                     0.0   
4836               0.0              0.0                     1.0   
8064               0.0              0.0                     0.0   
4418               0.0              0.0                     0.0   
1784               0.0              0.0                     1.0   

      Item_Type_Fruits and Vegetables  Item_Type_Hard Drinks  \
7503                              0.0                    0.0   
2957                              0.0                    0.0   
7031                              0.0                    0.0   
1084                              0.0                    0.0   
856                               0.0                    0.0   
...                               ...                    ...   
4753                              0.0                    0.0   
4836                              0.0                    0.0   
8064                              1.0                    0.0   
4418                              0.0                    0.0   
1784                              0.0                    0.0   

      Item_Type_Health and Hygiene  Item_Type_Household  Item_Type_Meat  \
7503                           0.0                  0.0             0.0   
2957                           1.0                  0.0             0.0   
7031                           0.0                  0.0             0.0   
1084                           0.0                  0.0             0.0   
856                            0.0                  0.0             1.0   
...                            ...                  ...             ...   
4753                           0.0                  0.0             0.0   
4836                           0.0                  0.0             0.0   
8064                           0.0                  0.0             0.0   
4418                           0.0                  1.0             0.0   
1784                           0.

Scale the Ordinal Features

In [ ]:
# instantiate scaler
scaler = StandardScaler()

In [ ]:
# fit scaler on training data
scaler.fit(X_train_ord_encoder)

StandardScaler()

In [ ]:
# Transform training data
X_train_ord_scaled = scaler.transform(X_train_ord_encoder)
# Transform testing data
X_test_ord_scaled = scaler.transform(X_test_ord_encoder)

Scale the categorical Features

In [ ]:
# instantiate scaler
scaler = StandardScaler()

In [ ]:
# fit scaler on training data
scaler.fit(X_train_cat_ohe)

StandardScaler()

In [ ]:
# Transform training data
X_train_cat_scaled = scaler.transform(X_train_cat_ohe)
# Transform testing data
X_test_cat_scaled = scaler.transform(X_test_cat_ohe)

Combine Training Data

In [ ]:
# re-combine the 3 dataframes of training data
X_train_processed = pd.concat([X_train_num_scaled,X_train_ord_scaled, X_train_cat_ohe], axis=1)
X_train_processed.info()
X_train_processed.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6392 entries, 4776 to 7270
Data columns (total 40 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Item_Weight                      6392 non-null   float64
 1   Item_Visibility                  6392 non-null   float64
 2   Item_MRP                         6392 non-null   float64
 3   Outlet_Establishment_Year        6392 non-null   float64
 4   Outlet_Size                      6392 non-null   float64
 5   Item_Fat_Content_Low Fat         6392 non-null   float64
 6   Item_Fat_Content_Regular         6392 non-null   float64
 7   Item_Type_Baking Goods           6392 non-null   float64
 8   Item_Type_Breads                 6392 non-null   float64
 9   Item_Type_Breakfast              6392 non-null   float64
 10  Item_Type_Canned                 6392 non-null   float64
 11  Item_Type_Dairy                  6392 non-null   float64
 12  Item_Type_Frozen 

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
4776     0.827485        -0.712775  1.828109                   1.327849   
7510     0.566644        -1.291052  0.603369                   1.327849   
5828    -0.121028         1.813319  0.244541                   0.136187   
5327    -1.158464        -1.004931 -0.952591                   0.732018   
4810     1.538870        -0.965484 -0.336460                   0.493686   

      Outlet_Size  Item_Fat_Content_Low Fat  Item_Fat_Content_Regular  \
4776     0.748125                       1.0                       0.0   
7510     0.748125                       0.0                       1.0   
5828     0.748125                       0.0                       1.0   
5327    -0.264370                       1.0                       0.0   
4810    -1.276865                       1.0                       0.0   

      Item_Type_Baking Goods  Item_Type_Breads  Item_Type_Breakfast  \
4776                     0.0               0.0                  0.0   
7510                     0.0               0.0                  0.0   
5828                     0.0               0.0                  0.0   
5327                     1.0               0.0                  0.0   
4810                     0.0               0.0                  0.0   

      Item_Type_Canned  Item_Type_Dairy  Item_Type_Frozen Foods  \
4776               0.0              0.0                     0.0   
7510               0.0              0.0                     0.0   
5828               0.0              0.0                     0.0   
5327               0.0              0.0                     0.0   
4810               0.0              0.0                     1.0   

      Item_Type_Fruits and Vegetables  Item_Type_Hard Drinks  \
4776                              0.0                    0.0   
7510                              0.0                    0.0   
5828                              0.0                    0.0   
5327                              0.0                    0.0   
4810                              0.0                    0.0   

      Item_Type_Health and Hygiene  Item_Type_Household  Item_Type_Meat  \
4776                           0.0                  1.0             0.0   
7510                           0.0                  0.0             0.0   
5828                           0.0                  0.0             1.0   
5327                           0.0                  0.0             0.0   
4810                           0.0                  0.0             0.0   

      Item_Type_Others  Item_Type_Seafood  Item_Type_Snack Foods  \
4776               0.0                0.0                    0.0   
7510               0.0                0.0                    1.0   
5828               0.0                0.0                    0.0   
5327               0.0                0.0                    0.0   
4810               0.0                0.0                    0.0   

      Item_Type_Soft Drinks  Item_Type_Starchy Foods  \
4776                    0.0                      0.0   
7510                    0.0                      0.0   
5828                    0.0                      0.0   
5327                    0.0                      0.0   
4810                    0.0                      0.0   

      Outlet_Identifier_OUT010  Outlet_Identifier_OUT013  \
4776                       0.0                       0.0   
7510                       0.0                       0.0   
5828                       0.0                       0.0   
5327                       0.0                       0.0   
4810                       0.0                       0.0   

      Outlet_Identifier_OUT017  Outlet_Identifier_OUT018  \
4776                       0.0                       1.0   
7510                       0.0                       1.0   
5828                       0.0                       0.0   
5327                       0.0                       0.0   
4810                       0.0             

Combine Testing Data

In [ ]:
# re-combine the 3 dataframes of testing data
X_test_processed = pd.concat([X_test_num_scaled, X_test_ord_scaled, X_test_cat_ohe], axis=1)
X_test_processed.info()
X_test_processed.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2131 entries, 7503 to 1784
Data columns (total 40 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Item_Weight                      2131 non-null   float64
 1   Item_Visibility                  2131 non-null   float64
 2   Item_MRP                         2131 non-null   float64
 3   Outlet_Establishment_Year        2131 non-null   float64
 4   Outlet_Size                      2131 non-null   float64
 5   Item_Fat_Content_Low Fat         2131 non-null   float64
 6   Item_Fat_Content_Regular         2131 non-null   float64
 7   Item_Type_Baking Goods           2131 non-null   float64
 8   Item_Type_Breads                 2131 non-null   float64
 9   Item_Type_Breakfast              2131 non-null   float64
 10  Item_Type_Canned                 2131 non-null   float64
 11  Item_Type_Dairy                  2131 non-null   float64
 12  Item_Type_Frozen 

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
7503     0.341372        -0.776646 -0.998816                  -1.293807   
2957    -1.169135         0.100317 -1.585194                  -0.102145   
7031     0.388798        -0.482994 -1.595784                   0.136187   
1084    -0.049889        -0.415440  0.506592                  -1.532139   
856     -0.632039        -1.047426  0.886725                   0.732018   

      Outlet_Size  Item_Fat_Content_Low Fat  Item_Fat_Content_Regular  \
7503     1.760619                       1.0                       0.0   
2957    -0.264370                       1.0                       0.0   
7031     0.748125                       0.0                       1.0   
1084     0.748125                       0.0                       1.0   
856     -0.264370                       0.0                       1.0   

      Item_Type_Baking Goods  Item_Type_Breads  Item_Type_Breakfast  \
7503                     0.0               0.0                  0.0   
2957                     0.0               0.0                  0.0   
7031                     0.0               0.0                  0.0   
1084                     0.0               0.0                  0.0   
856                      0.0               0.0                  0.0   

      Item_Type_Canned  Item_Type_Dairy  Item_Type_Frozen Foods  \
7503               0.0              0.0                     1.0   
2957               0.0              0.0                     0.0   
7031               1.0              0.0                     0.0   
1084               0.0              0.0                     0.0   
856                0.0              0.0                     0.0   

      Item_Type_Fruits and Vegetables  Item_Type_Hard Drinks  \
7503                              0.0                    0.0   
2957                              0.0                    0.0   
7031                              0.0                    0.0   
1084                              0.0                    0.0   
856                               0.0                    0.0   

      Item_Type_Health and Hygiene  Item_Type_Household  Item_Type_Meat  \
7503                           0.0                  0.0             0.0   
2957                           1.0                  0.0             0.0   
7031                           0.0                  0.0             0.0   
1084                           0.0                  0.0             0.0   
856                            0.0                  0.0             1.0   

      Item_Type_Others  Item_Type_Seafood  Item_Type_Snack Foods  \
7503               0.0                0.0                    0.0   
2957               0.0                0.0                    0.0   
7031               0.0                0.0                    0.0   
1084               0.0                0.0                    0.0   
856                0.0                0.0                    0.0   

      Item_Type_Soft Drinks  Item_Type_Starchy Foods  \
7503                    0.0                      0.0   
2957                    0.0                      0.0   
7031                    0.0                      0.0   
1084                    1.0                      0.0   
856                     0.0                      0.0   

      Outlet_Identifier_OUT010  Outlet_Identifier_OUT013  \
7503                       0.0                       1.0   
2957                       0.0                       0.0   
7031                       0.0                       0.0   
1084                       0.0                       0.0   
856                        0.0                       0.0   

      Outlet_Identifier_OUT017  Outlet_Identifier_OUT018  \
7503                       0.0                       0.0   
2957                       0.0                       0.0   
7031                       0.0                       0.0   
1084                       0.0                       0.0   
856                        0.0             

Make sure your imputation of missing values occurs after the train test split using SimpleImputer.

In [ ]:
X_train_processed.isna().sum().sum()

0

In [ ]:
X_test_processed.isna().sum().sum()

0

# FINAL PART

1. Build a linear regression model to predict sales.

* Build a linear regression model

In [ ]:
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg

LinearRegression()

In [ ]:
# Fit the model on the training data
lin_reg.fit(X_train_processed, y_train)

LinearRegression()

In [ ]:
# Get predictions for the training data
y_predictions_train = lin_reg.predict(X_train_processed)
# Get predictions for the testing data
y_predictions_test = lin_reg.predict(X_test_processed)

* Evaluate the performance of your model based on r^2

In [ ]:
from sklearn.metrics import r2_score
train_r2 = r2_score(y_train, y_predictions_train)
test_r2 = r2_score(y_test, y_predictions_test)
print(f'Model Training R2: {train_r2:.2f}')
print(f'Model Testing R2: {test_r2:.2f}')

Model Training R2: 0.56
Model Testing R2: 0.57


* Evaluate the performance of your model based on rmse.

In [ ]:
from sklearn.metrics import mean_squared_error
train_RMSE = mean_squared_error(y_train, y_predictions_train, squared=False)
test_RMSE = mean_squared_error(y_test, y_predictions_test, squared=False)
print(f'Model Training RMSE: {train_RMSE:,.2f}')
print(f'Model Testing RMSE: {test_RMSE:,.2f}')

Model Training RMSE: 1,139.10
Model Testing RMSE: 1,092.81


2. Your second task is to build a regression tree model to predict sales.

* Build a simple regression tree model

In [ ]:
from sklearn.tree import DecisionTreeRegressor
# Make an instance of the model
dec_tree = DecisionTreeRegressor(random_state = 42)
# Training the Model on the data
dec_tree.fit(X_train_processed, y_train)
# Make predictions for training and testing data
train_preds_dec = dec_tree.predict(X_train_processed)
test_preds_dec = dec_tree.predict(X_test_processed)

* Compare the performance of your model based on r^2.

In [ ]:
train_score = r2_score(y_train, train_preds_dec)
test_score = r2_score(y_test, test_preds_dec)
print (f'Training r2 is {train_score:.2f}.')
print (f'Testing r2 is {test_score:.2f}.')

Training r2 is 1.00.
Testing r2 is 0.19.


* Compare the performance of your model based on rmse.

In [ ]:
train_RMSE_dec = mean_squared_error(y_train, train_preds_dec,squared=False)
test_RMSE_dec = mean_squared_error(y_test, test_preds_dec,squared=False)
print (f'Training RMSE is {train_RMSE_dec:.2f}.')
print (f'Testing RMSE is {test_RMSE_dec:.2f}.')

Training RMSE is 0.00.
Testing RMSE is 1492.52.


3. You now have tried 2 different models on your data set. You need to determine which model to implement.
* I recommend the first model because of its regression metrics are more reliable for example R2 in the first model(linear regression model) is the same almost for both the training data and the testing data , our model can account for about 57% of the variation in y_test using the features in X_test.But for the second model there is big difference between trainig R2 which is equal to 1.0 and testing R2 which is so lower (0.19) ,this model can account for about 19% of the variation in y_test using the features in X_test.
Also we note that RMSE in the second model (1492.52) is significantly larger than RMSE in the first model (1,092.81)